### PR TITLE
Extract more accurate glyph heights from type3 fonts

### LINKED ISSUE #5896: Text height returned by getTextContent is way off
I am having an issue extracting text from [this PDF](http://www.dfki.uni-kl.de/~kieni/publications/spie98.pdf). My goal is to extract the text bounding-boxes via getTextContent. The problem is, the height of the extracted text chunks are off by a large margin. The height I am getting (for all texts) is `0.12`, while the actual height looks closer to `12.0`.

In the native adobe-reader, the text bounding boxes are correct, as evidenced by highlighting the text. Furthermore, the text height is rendered correctly by PDFJS (the text is indeed being rendered, its not an image), so I don't believe the PDF is providing incorrect info.

After some digging around, it looks like the issue might be related to the textMatrix. For the all the texts on the page, the TM scaling is `[0.12, 0, 0, -0.12, tx, ty]`, the CTM is `[1, 0, 0, 1, 0, 0]`, while the fontSize is `1`. Is this information being wrongly parsed/scaled? How does the renderer get it right?

The issue is experienced both in Chrome and Firefox.

Below is a screenshot of the text bounding-boxes drawn over the rendered page.
screenshot

### PR DESCRIPTION
**Pull request by stri8ed**
Unlike other font-types, type-3 fonts define a custom transformation from glyph-space to text-space, via the `FontMatrix`. As such, we cannot assume the glyph's to have a height of `1 unit` (as is done currently). By utilizing the `FontBBox` (which is a required property of type-3 fonts), we can get a good estimate of the glyph height.

From the spec:

> The glyph coordinate system is the space in which an individual character’s glyph is defined. All path coordinates and metrics shall be interpreted in glyph space. For all font types except Type 3, the units of glyph space are one-thousandth of a unit of text space; for a Type 3 font, the transformation from glyph space to text space shall be defined by a font matrix specified in an explicit FontMatrix entry in the font.

And

> A font defines the glyphs at one standard size. This standard is arranged so that the nominal height of tightly spaced lines of text is 1 unit

Improvements can be verified by text-selection in these PDF's:

http://www.dfki.uni-kl.de/~kieni/publications/spie98.pdf

http://neo.lcc.uma.es/Articles/WRH98.pdf

Fixes #5896.

### CODE REVIEW

**Comment by timvandermeij** [comment]
Nice work! I can confirm that this fixes #3360 too. I wonder if the text failures aren't expected as we now respect the `bbox` property, but I'm sure that you know a lot more about that than I do!

File: src/core/evaluator.js [hunk 1]
```diff
@@ -956,6 +956,13 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
           var tsm = [textState.fontSize * textState.textHScale, 0,
                      0, textState.fontSize,
                      0, textState.textRise];
+                 
+          if(font.isType3Font) {
```
Comment (timvandermeij) [inline src/core/evaluator.js]: Nit: add a space after `if`: `if (font.isType3Font) {`

**Comment by stri8ed** [comment]
It seems the reference snapshots are generated via the existing `buildTextGeometry` method. Hence,  the two reference tests failing, is to be expected, considering that the patch fixes, and thus changes the text geometries. I can confirm that when comparing the extracted text bounding-boxes against the glyph's rendered via `renderPage`, the accuracy is considerably improved.

**Comment by stri8ed** [comment]
I added a check for heights of zero, as per the spec:
>If all four elements of the rectangle are zero, a conforming reader shall make no assumptions about glyph sizes based on the font bounding box.

**Comment by Snuffleupagus** [comment]
I still believe that this patch actually causes *too* much movement for some Type3 fonts.
When comparing the `textLayer` rendering of http://www.pnml.kz/pdfexample/v14c0009753.20140818.ru.pdf, taken from issue #5421 (which is the basis for one of the failing tests), it's quite clear that with this patch the `textLayer` glyphs are too tall:
 - Current master
master

 - This patch
patch

File: src/core/fonts.js [hunk 2]
```diff
@@ -2438,6 +2438,7 @@ var Font = (function FontClosure() {
     this.ascent = properties.ascent / PDF_GLYPH_SPACE_UNITS;
     this.descent = properties.descent / PDF_GLYPH_SPACE_UNITS;
     this.fontMatrix = properties.fontMatrix;
+    this.bbox = properties.bbox;
```
Comment (Snuffleupagus) [inline src/core/fonts.js]: Nit: this line seems unrelated to the patch.

**Comment by Snuffleupagus** [comment]
Looking at the difference between the Type3 files in the test suite and those that this PR tries to fix, there's a marked difference between the `FontMatrix` entries.
In the files in the test suite, it's `[0.001, 0, 0, 0.001, 0, 0]` (i.e. identical to `FONT_IDENTITY_MATRIX`), and for the files referenced here it's `[1, 0, 0, -1, 0, 0]`.

Hence I'm wondering if we should perhaps only use the `FontBBox` values for Type3 fonts when `FontMatrix !== FONT_IDENTITY_MATRIX`.

File: src/core/evaluator.js [hunk 3]
```diff
@@ -956,6 +956,15 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
           var tsm = [textState.fontSize * textState.textHScale, 0,
                      0, textState.fontSize,
                      0, textState.textRise];
+
```
Comment (Snuffleupagus) [inline src/core/evaluator.js]: There's a bunch of "spaces" left on this line, please remove them.

File: src/core/evaluator.js [hunk 4]
```diff
@@ -956,6 +956,15 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
                      0, textState.textRise];
+                 
+          if (font.isType3Font) {
+            var glyphHeight = font.bbox[3] - font.bbox[1];
+            if (glyphHeight > 0) {
+              glyphHeight = glyphHeight * textState.fontMatrix[3];
+              tsm[3] *= glyphHeight;
+            }
+          }
+
```
Comment (Snuffleupagus) [inline src/core/evaluator.js]: There's a bunch of "spaces" left on this line, please remove them.

**Comment by stri8ed** [comment]
`FontBBox` gives us the height of the tallest glyph contained in the font. So naturally, this will not always be super accurate. By not using the `FontBBox`, we are assuming the glphy's to have a height of `1`, which seems to work in many cases. I suppose we could look up the actual height of the glpyhs as defined in the `CharProcs`, but this will cost performance, and the text-chunk will take on the height of the first glyph (unless we get heights foreach glyph, and take the max). 

Even if the FontMatrix is `FONT_IDENTITY_MATRIX`, this is no guarantee that the glyph heights will be close to `1`. Unless I am mistaken?

**Comment by Snuffleupagus** [comment]
@stri8ed As I've shown in https://github.com/mozilla/pdf.js/pull/5898#issuecomment-89062730, there are cases where this patch *would* regress the current behaviour. And unless we have a *very* compelling reason for regressing something, we always strive to avoid it.

To correctly handle various kinds of strange PDF data found in the wild, the code-base thus contain a fair number of heuristics to try and account for that.
Based on a quick look at a couple of PDF files with Type3 fonts, what I was suggesting is simply to look into using a few heuristics to avoid unnecessary regressions (e.g. only use the `FontBBox` data when `textState.fontSize === 1 && textState.fontMatrix !== FONT_IDENTITY_MATRIX`).

**Comment by stri8ed** [comment]
Fair enough. Added that, and removed the extra spaces.

**Comment by timvandermeij** [comment]
@Snuffleupagus Would you be comfortable doing the final review for this patch? If so, feel free to change the assignee.

**Comment by timvandermeij** [comment]
Thank you for the patch!
